### DISCUSSION TITLE
local maximum slower in lasR

### DISCUSSION THREAD

**Discussion by wiesehahn** [r-lidar/lasR#38] [General]
I just had a look at the benchmark vignette and wondered that everything seems to be faster in lasR 0.5. compared to lidR except the local maximum function, which is apperently slower in single and multicore processing. Is this expected?

https://r-lidar.github.io/lasR/articles/benchmarks.html#result-4

<img src="https://r-lidar.github.io/lasR/articles/benchmarks_files/figure-html/unnamed-chunk-7-1.png" width="350"> <img src="https://r-lidar.github.io/lasR/articles/benchmarks_files/figure-html/unnamed-chunk-7-2.png" width="350">

**Comment by Jean-Romain** [comment]
Yes, it is not surprising. It was originally 20 times faster than `lidR` but then I implemented the same optimization in [`lidR v4.1.1 `](https://github.com/r-lidar/lidR/blob/master/NEWS.md) to make `lidR` 20x faster with fixed windows size.

Now, `lidR` and `lasR` are supposed to be on par. Why `lasR` is slightly slower? I don't know exactly but the most relevant explanation comes from the memory layout. The code is strictly the same but the memory layout is very different. `lasR` keeps the memory tight and optimized and it should be (did not ran proper benchmarks) a little bit slower to read the point cloud in memory and make spatial queries in `lasR`. Moreover `lasR` is capable of assigning a unique idea from 1 to `n` to each tree in parallel, while `lidR` is not. There is probably an overhead for this feature.

As you can see in this benchmark this saves 2 GB of memory at the cost a 5 seconds. 2 GB saved means more memory available for mores cores which is significant. The additional 5 seconds are non significant if you consider you can pipe the operations and divide the overall computation by 2,3, or or 10 compared to `lidR` or `LAStools` with the piping system on complex pipelines.

So all in all, it is not a big deal and I have no concern with that.

**Comment by Jean-Romain** [comment]
Ho, actually I think I have a better anwser. The difference likely comes from the writing of the maxima in the output vector file. It is slow I needs improvement.
